\documentclass[11pt]{article}
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
    \usepackage{xltxtra,xunicode}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Mapping=tex-text,Scale=MatchLowercase}
  \newcommand{\euro}{â‚¬}
\fi
% use microtype if available
\IfFileExists{microtype.sty}{\usepackage{microtype}}{}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{{#1}}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{{#1}}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{{#1}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{{#1}}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\RegionMarkerTok}[1]{{#1}}
\newcommand{\ErrorTok}[1]{\textbf{{#1}}}
\newcommand{\NormalTok}[1]{{#1}}
\usepackage{graphicx}
% Redefine \includegraphics so that, unless explicit options are
% given, the image width will not exceed the width of the page.
% Images get their normal width if they fit onto the page, but
% are scaled down if they would overflow the margins.
\makeatletter
\def\ScaleIfNeeded{%
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother
\let\Oldincludegraphics\includegraphics
{%
 \catcode`\@=11\relax%
 \gdef\includegraphics{\@ifnextchar[{\Oldincludegraphics}{\Oldincludegraphics[width=\ScaleIfNeeded]}}%
}%
\ifxetex
  \usepackage[setpagesize=false, % page size defined by xetex
              unicode=false, % unicode breaks when used with xetex
              xetex]{hyperref}
\else
  \usepackage[unicode=true]{hyperref}
\fi
\hypersetup{breaklinks=true,
            bookmarks=true,
            pdfauthor={},
            pdftitle={},
            colorlinks=true,
            citecolor=blue,
            urlcolor=blue,
            linkcolor=magenta,
            pdfborder={0 0 0}}
\urlstyle{same}  % don't use monospace font for urls
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
%\setcounter{secnumdepth}{0}
\usepackage{fullpage}
\usepackage{framed}
\usepackage{mathtools}
\usepackage[osf]{mathpazo} % palatino
\usepackage{float}

\begin{document}

\title{Introduction to Phylogenetic Comparative Methods in R: Models of evolution}
\author{Natalie Cooper (ncooper@tcd.ie) and Graham Slater (gslater@si.edu)}
\date{}
\maketitle

This problem set aims to show you how to use R to answer phylogenetic 
comparative questions. By the end of this problem set you should be able to:

\begin{enumerate}
\item Read your data and phylogeny into R
\item View and manipulate your data and phylogeny
\item Match taxa in your data with those in your phylogeny
\item Fit simple models of evolution to continuous data
\item Graham's magical OUwie stuff
\end{enumerate}

We will be using the evolution of primate life-history variables as an
example. These data come from the PanTHERIA database (Jones \textit{et
al}. 2009) and 10kTrees (Arnold \textit{et al}. 2010). Note that this is
an old version of 10kTrees, so if you want to use it in your research
please download the newest version.

Throughout, R code will be in shaded boxes:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{install.packages}\NormalTok{(}\StringTok{"ape"}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

R output will be preceded by \texttt{\#\#} and important comments will be in boxes:

\begin{framed}

The handout assumes that people have little or no experience with R so 
more experienced users may want to skip quickly through the first 
few sections. \textbf{To get the analyses from section 3 onwards to work you 
will need to have completed sections 1.1, 1.3, 1.4, 2.1 (reading in data only), 
2.2 (reading in phylogeny only) and 2.3.} Other sections are included to show 
you some of the other things R can do and to help you with problems you are 
likely to encounter when using your own data. 
\end{framed}

\newpage{}
\section{Preparations}

\begin{framed}
Note that many things in R can be done in multiple ways. You should
choose the methods you feel most comfortable with, and do not panic if
someone is doing the same analyses as you in a different way! This
workshop will be full of different ways to do things.
\end{framed}

\subsection{Downloading the data and finding the path of your folder}

First you need to download all the files for this problem set into a
folder somewhere on your computer, let's call it ``RAnalyses2014'' and
pop it on the Desktop. We will use this folder throughout the problem
set. You'll need to know what the \textbf{path} of the folder is. For
example on Natalie's Windows machine, the path is:

\begin{snugshade}
\texttt{C:/Users/Natalie/Desktop/RAnalyses2014}
\end{snugshade}

The path is really easy to find in a Windows machine, just click on the
address bar of the folder and the whole path will appear.

\begin{framed}
In Windows, paths usually include \textbackslash{} but R
can't read these. It's easy to fix in your R code, just change any \textbackslash{} in
the path to / or \textbackslash{}\textbackslash{}.
\end{framed}

On Natalie's Mac the path is:

\begin{snugshade}
\texttt{\textasciitilde{}/Desktop/RAnalyses2014}
\end{snugshade}

It's a bit trickier to find the path on a Mac, so ask if you need help. 
Note that the tilde \textasciitilde{} is a shorthand for /Users/Natalie. 

\subsection{Using a text editor}

Next, open a text editor. R has an inbuilt editor that works pretty
well, but NotePad and TextEdit are fine too. However, in the future we
\textbf{highly} recommend using something that will highlight code for
you. Natalie's personal favorite is Sublime Text 2, because you can also use
it for any other kind of text editing like LaTeX, html etc.

You should type (or copy and paste) your code into the text editor, edit
it until you think it'll work, and then paste it into R's console
window. Saving the text file lets you keep a record of the code you
used, which can be a great timesaver if you want to use it again,
especially as you know this code will work! It is also \emph{essential} to
keep \textbf{all} code used to produce the results of analyses and figures
in your publications.

\begin{framed}
You can cut and paste code from this handout into your text editor, or 
straight into R. You don't need to retype everything!
\end{framed}

If you want to add comments to the file (i.e., notes to remind yourself
what the code is doing), put a hash/pound sign (\#) in front of the
comment.

\begin{snugshade}
\texttt{\# Comments are ignored by R but can remind you what the code is doing.\\}
\texttt{\# You need a hash sign at the start of each line of a comment.}
\end{snugshade}

\subsection{Installing packages in R}\label{installing-packages}

To run comparative analyses (or any specialized analysis) in R, you need to 
download one or more additional packages from the basic R installation. 
For this problem set you will need to install the following packages: 
\texttt{ape}, \texttt{geiger} and \texttt{OUwie}. 
To install the package \texttt{ape}:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{install.packages}\NormalTok{(}\StringTok{"ape"}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Now install \texttt{geiger} and \texttt{OUwie}.

\subsection{Loading packages in R}

You've installed the packages but they don't automatically get loaded
into your R session. Instead you need to tell R to load them \textbf{every
time} you start a new R session and want to use functions from these
packages. To load the package \texttt{ape} into your current R session:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ape)}
\end{Highlighting}
\end{snugshade}

Don't forget to load \texttt{geiger} and \texttt{OUwie} too!

\subsection{Citing packages in R}\label{citing-packages}

Note that all these extra packages take lots of work to write so
make sure to cite them properly if you use them. You can get
the citation of a package by typing: 

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{citation}\NormalTok{(package = }\StringTok{"ape"}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

You should also cite R. You can get the correct citation for the version
you are using as follows:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{citation}\NormalTok{()}
\end{Highlighting}
\end{snugshade}

Make sure to include version numbers in your citations for reproducibility purposes.

\section{Reading your data and phylogeny into R}
\subsection{Reading data into R}

Next we need to load the data we are going to use for the analysis. R
can read files in lots of formats, including comma-delimited and
tab-delimited files. Excel (and many other applications) can output
files in this format (it's an option in the ``Save As'' dialog box
under the ``File'' menu). To save time we have given you a tab-delimited
text file called ``Primatedata.txt'' which we are going to use. Load
these data as follows. 

\begin{framed}
You will need to replace MYPATH with the name of the path to the folder 
containing the data and tree on your computer.
\end{framed}

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatedata <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\StringTok{"MYPATH/Primatedata.txt"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }
                          \DataTypeTok{header =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Note that \texttt{sep = "\textbackslash{}t"} indicates that you have a tab-delimited file, 
\texttt{sep = ","}  would indicate a comma-delimited csv file. You can also use
\texttt{read.delim} for tab delimited files or \texttt{read.csv} for comma delimited
files. \texttt{header = TRUE}, indicates that the first line of the data contains
column headings.

This is a good point to note that unless you \textbf{tell} R you want to
do something, it won't do it automatically. So here if you successfully
entered the data, R won't give you any indication that it worked.
Instead you need to specifically ask R to look at the data.

We can look at the data by typing:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{str}\NormalTok{(primatedata)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## 'data.frame':    77 obs. of  8 variables:
##  $ Order          : Factor w/ 1 level "Primates": 1 1 1 ...
##  $ Family         : Factor w/ 15 levels "Aotidae","Atelidae",...
##  $ Binomial       : Factor w/ 77 levels "Alouatta palliata",...
##  $ AdultBodyMass_g: num  6692 7582 8697 958 558 ...
##  $ GestationLen_d : num  138 226 228 164 154 ...
##  $ HomeRange_km2  : num  2.28 0.73 1.36 0.02 0.32 0.02 ...
##  $ MaxLongevity_m : num  336 328 454 304 215 ...
##  $ SocialGroupSize: num  14.5 42 20 2.95 6.85 ...
\end{verbatim}

This shows the structure of the data frame (this can be a really useful
command when you have a big data file). It also tells you what kind of
variables R thinks you have (characters, integers, numeric, factors
etc.). Some R functions need the data to be certain kinds of variables
so it's useful to check this.

As you can see, the data contains the following variables: Order,
Family, Binomial, AdultBodyMass\_g, GestationLen\_d, HomeRange\_km2,
MaxLongevity\_m, and SocialGroupSize.

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(primatedata)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
##      Order      Family           Binomial AdultBodyMass_g GestationLen_d
## 1 Primates    Atelidae   Ateles belzebuth          6692.4          138.2
## 2 Primates    Atelidae   Ateles geoffroyi          7582.4          226.4
## 3 Primates    Atelidae    Ateles paniscus          8697.2          228.2
## 4 Primates Pitheciidae  Callicebus moloch           958.1          164.0
## 5 Primates     Cebidae  Callimico goeldii           558.0          154.0
## 6 Primates     Cebidae Callithrix jacchus           290.2          144.0
##   HomeRange_km2 MaxLongevity_m SocialGroupSize
## 1          2.28          336.0           14.50
## 2          0.73          327.6           42.00
## 3          1.36          453.6           20.00
## 4          0.02          303.6            2.95
## 5          0.32          214.8            6.85
## 6          0.02          201.6            8.55
\end{verbatim}

This gives you the first few rows of data along with the column
headings.

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{names}\NormalTok{(primatedata)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## [1] "Order"           "Family"          "Binomial"        "AdultBodyMass_g"
## [5] "GestationLen_d"  "HomeRange_km2"   "MaxLongevity_m"  "SocialGroupSize"
\end{verbatim}

This gives you the names of the columns.

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatedata}
\end{Highlighting}
\end{snugshade}

This will print out all of the data!

\subsection{Reading and displaying your phylogeny in R}

To load a tree you need either the function \texttt{read.tree} or \texttt{read.nexus}.
\texttt{read.tree} can deal with a number of different types of data (including
DNA) whereas \texttt{read.nexus} reads NEXUS files. We will use a NEXUS file of
the consensus tree from 10kTrees.

\begin{framed}
You will need to replace MYPATH with the name of the path to the folder containing the data and tree on your computer.
\end{framed}

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree <-}\StringTok{ }\KeywordTok{read.nexus}\NormalTok{(}\StringTok{"MYPATH/consensusTree_10kTrees_Version2.nex"}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Let's examine the tree by typing:

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## Phylogenetic tree with 226 tips and 221 internal nodes.
## 
## Tip labels:
##  Allenopithecus_nigroviridis, Cercopithecus_ascanius, 
Cercopithecus_cephus, Cercopithecus_cephus_cephus, ...
## 
## Rooted; includes branch lengths.
\end{verbatim}

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{str}\NormalTok{(primatetree)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## List of 4
##  $ edge       : int [1:446, 1:2] 227 228 229 230 231 232 ...
##  $ edge.length: num [1:446] 4.95 17.69 19.65 8.12 4.82 ...
##  $ Nnode      : int 221
##  $ tip.label  : chr [1:226] "Allenopithecus_nigroviridis" ...
##  - attr(*, "class")= chr "phylo"
##  - attr(*, "order")= chr "cladewise"
\end{verbatim}

\texttt{primatetree} is a fully resolved tree with branch lengths. There are 226
species and 221 internal nodes. We can plot the tree by using the \texttt{plot}
function of \texttt{ape}:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(primatetree}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Note that the tree has too many species for us to read the species names. You can make the tip labels smaller, but that doesnâ€™t help much here:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(primatetree, }\DataTypeTok{cex =} \FloatTok{0.5}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Alternatively you can zoom into different sections of the tree that youâ€™re interested in:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{zoom}\NormalTok{(primatetree, }\KeywordTok{list}\NormalTok{(}\KeywordTok{grep}\NormalTok{(}\StringTok{"Cercopithecus"}\NormalTok{, primatetree\$tip.label)), }
      \DataTypeTok{subtree =} \OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

This just gives you the tree for \textit{Cercopithecus} species but you can also see how the species fit into the rest of the tree using:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{zoom}\NormalTok{(primatetree, }\KeywordTok{list}\NormalTok{(}\KeywordTok{grep}\NormalTok{(}\StringTok{"Cercopithecus"}\NormalTok{, primatetree\$tip.label)), }
      \DataTypeTok{subtree =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Note that \texttt{zoom} automatically sets the plotting window to display two plots at once. To reset this to one plot only use:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{snugshade}

You can also remove species from the tree very easily using the \texttt{ape} function \texttt{drop.tip}:

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree2 <-}\StringTok{ }\KeywordTok{drop.tip}\NormalTok{(primatetree, } \StringTok{"Aotus_azarae_infulatus"}\NormalTok{)}\\

\KeywordTok{str}\NormalTok{(primatetree2)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## List of 4
##  $ edge       : int [1:444, 1:2] 226 227 228 229 ...
##  $ edge.length: num [1:444] 4.95 17.69 19.65 8.12 ...
##  $ Nnode      : int 220
##  $ tip.label  : chr [1:225] "Allenopithecus_nigroviridis" ...
##  - attr(*, "class")= chr "phylo"
##  - attr(*, "order")= chr "cladewise"
\end{verbatim}

To remove more than one species see below. To get further options for the plotting of phylogenies:

\begin{snugshade}
\begin{Highlighting}[]
\DataTypeTok{?}\NormalTok{(plot.phylo)}
\end{Highlighting}
\end{snugshade}

Note that although you can use \texttt{plot} to plot the phylogeny, you need to specify \texttt{plot.phylo} to find out the options for plotting trees. You can change the style of the tree (type), the color of the branches and tips (edge.color, tip.color). Hereâ€™s an fun* example! *Definitions of fun may vary.

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{))}

\KeywordTok{plot}\NormalTok{(primatetree, }\DataTypeTok{type =} \StringTok{"fan"}\NormalTok{, }\DataTypeTok{edge.color =} \StringTok{"deeppink"}\NormalTok{, }\DataTypeTok{tip.color =} 
      \StringTok{rainbow(8)}\NormalTok{, }\DataTypeTok{cex =} \FloatTok{0.5}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

You can also add a timescale to your tree:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(primatetree)}

\KeywordTok{axisPhylo}\NormalTok{()}
\end{Highlighting}
\end{snugshade}

For many more options for tree plotting check out Liam Revell's \texttt{phytools} package.

\subsection{Making your phylogeny binary and rooting your phylogeny in R}

Most R functions require your tree to be dichotomous, i.e. to have no polytomies. To check whether your tree is dichotomous use \texttt{is.binary.tree}. If this is FALSE, use \texttt{multi2di} to make the tree dichotomous. This function works by randomly resolving polytomies with zero-length branches.

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{is.binary.tree}\NormalTok{(primatetree)}\CommentTok{# we want this to be TRUE}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## [1] FALSE
\end{verbatim}

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree <-}\StringTok{ }\KeywordTok{multi2di}\NormalTok{(primatetree)}
\end{Highlighting}
\end{snugshade}

Most functions also require the tree to be rooted, i.e., to have one taxon designated as the outgroup. Our tree is rooted but if you wanted to change the root, or root an unrooted tree use \texttt{root}. Note that here we've just chosen a random species (\textit{Saimiri sciureus}) to be the root.

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree.reroot <-}\StringTok{ }\KeywordTok{root}\NormalTok{(primatetree, }\StringTok{"Saimiri_sciureus"}\NormalTok{)  }

\KeywordTok{plot}\NormalTok{(primatetree.reroot)}
\end{Highlighting}
\end{snugshade}

\subsection{Matching species names in your data and phylogeny in R}

\subsubsection{Species names with spaces} 
Species names in the tree cannot contain spaces so they are generally written as Genus\_species (the gap between the genus name and species name replaced by \_). If the species names in the data are written as Genus species with a space, then you will have to replace the spaces with \_ so that they match up with the species names in the tree. You can do this as follows:

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatedata\$Binomial <-}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\StringTok{" "}\NormalTok{, }\StringTok{"_"}\NormalTok{, primatedata\$Binomial)}
\end{Highlighting}
\end{snugshade}

\texttt{gsub} means \textbf{g}eneral \textbf{sub}stitution. It replaces any instance of the first item (here itâ€™s a space) with the second item (\_) but only in the variable you tell it to (\texttt{primatedata\$Binomial}).

\subsubsection{Species names = row names}

A number of R functions require that species names are the row names of the data. This is really easy to fix:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{row.names}\NormalTok{(primatedata) <-}\StringTok{ }\NormalTok{primatedata\$Binomial}
\end{Highlighting}
\end{snugshade}

\subsubsection{Mismatches between species in your data and phylogeny}
Often you will have data for species which are not in your phylogeny and/or species in your phylogeny which are not in your data. Some functions in R can deal with this, others will produce an error telling you the tree and data do not match (e.g., most \texttt{ape} functions). It's useful to know how to deal with this so we have provided code below.

\begin{framed}
Note that many R functions match up the species names in the tree and data for you before you run any analyses. However, these functions are only as good as their inputs. If you have even slightly misspelled a species name in the tree or the data it will automatically be dropped from the analyses. It is therefore \textbf{very important} to check this before running an analysis.
\end{framed}

\subsubsection{Species in the phylogeny but not in the data}
These are easy to identify using \texttt{setdiff}:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{setdiff}\NormalTok{(primatetree\$tip.label, primatedata\$Binomial)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
##   [1] "Allenopithecus_nigroviridis"                  
##   [2] "Cercopithecus_cephus_cephus"                  
##   [3] "Cercopithecus_cephus_ngottoensis"             
##   [4] "Cercopithecus_diana" 
etc.
\end{verbatim}

\texttt{setdiff} tells you what is found in the first list, but not in the second. Here, it has listed the species that are found in the tree but \textbf{not} in the data. We can then use \texttt{setdiff} with \texttt{drop.tip} to prune the tree to just the species we have data for. 

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{primatetree2 <-}\StringTok{ }\KeywordTok{drop.tip}\NormalTok{(primatetree, }\KeywordTok{setdiff}\NormalTok{(primatetree\$tip.label, }
                          \NormalTok{primatedata\$Binomial))}
\end{Highlighting}
\end{snugshade}

Note that you need to list the species which you do \textbf{not} want to select and then drop them from the tree instead of selecting the species you want.

\subsubsection{Species in the data but not in the phylogeny}
Again these are easy to identify using \texttt{setdiff}, just swap the inputs around:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{setdiff}\NormalTok{(primatedata\$Binomial, primatetree\$tip.label)}
\end{Highlighting}
\end{snugshade}

\begin{verbatim}
## character(0)
\end{verbatim}

In this case we don't have any species in the tree missing from the data. However, if you do, to remove species from the data which are not in the tree you can use \texttt{match} and \texttt{subset}:

\begin{snugshade}
\begin{Highlighting}[]
\NormalTok{matches <-}\StringTok{ }\KeywordTok{match}\NormalTok{(primatedata\$Binomial, primatetree2\$tip.label, }\DataTypeTok{nomatch =} \DecValTok{0}\NormalTok{)}

\NormalTok{primatedata2 <-}\StringTok{ }\KeywordTok{subset}\NormalTok{(primatedata, matches !=}\StringTok{ }\DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{snugshade}

Remember \texttt{!=} means ``does not equal''. So this line of code only selects species which do appear in the tree, i.e. their value from \texttt{matches} is not 0. 

Always check this has worked as expected by checking the data and the phylogeny. In the first instance you can just use \texttt{str} to make sure you have the expected number of species in each:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{str}\NormalTok{(primatedata2)}

\KeywordTok{str}\NormalTok{(primatetree2)}
\end{Highlighting}
\end{snugshade}

\section{Fitting simple models of evolution to continuous data}

For fitting models of evolution to continuous data we will use the \texttt{fitContinuous} function in the R package \texttt{geiger}.

\texttt{fitContinuous} is a likelihood based method, so the output will give the maximum likelihood (ML) estimates of the parameters. Bayesian methods are becoming preferred for these kinds of analyses and \texttt{fitContinuousMCMC} will perform these analyses. However, due to time constraints (and a need to accompany it with a primer on Bayesian statistics!) we will not cover this function.

As an example, letâ€™s look at the evolution of log(body size) in Primates. Weâ€™ll fit three evolutionary models â€“ the Brownian motion (BM) model, the Ornstein-Uhlenbeck (OU) model and the Early Burst (EB) model. \texttt{fitContinuous} can also fit several other models. For more details look at the help file by typing:

\begin{snugshade}
\begin{Highlighting}[]
\KeywordTok{?}\NormalTok{fitContinuous}
\end{Highlighting}
\end{snugshade}

\subsection{Fitting Brownian motion (BM) models}

\subsubsection{Model description}

In the Brownian motion model (Felsenstein 1973), a trait X evolves at random at a rate %ðœŽ:
%dX(t)=ÏƒdW(t)            eqn (1)
where W(t) is a white noise function and is a random variate drawn from a normal distribution with mean 0 and variance %ðœŽ2. 
This model assumes that there is no overall drift in the direction of evolution (hence the expectation of W(t) is zero) and that the rate of evolution is constant. The model has two parameters, %ðœŽ 
and the state of the root at time zero, X(0). The Brownian model predicts after a time T the variance in trait value Xi for species i is:
%var(X_i )=Ïƒ^2 T                     eqn (2)
and the covariance in traits for species i and j is:
%cov(X_i,X_j )=Ïƒ^2 t_ij                           eqn (3)
where tij is the shared evolutionary pathway for species i and j, i.e. the time at which they last shared a common ancestor.  

Put in more simple terms, the Brownian motion model assumes the correlation structure among trait values is proportional to the extent of shared ancestry for pairs of species. This means that close relatives will be more similar in their trait values than more distant relatives. It also means that variance in the trait will increase (linearly) with time. 

\subsubsection{Model parameters}

\texttt{fitContinuous} estimates sigma squared, the Brownian rate parameter. This can be used as a proxy for the rate of evolution.\\

Default bounds on the Brownian rate parameter are sigsq=c(min=0,max=Inf). 

\subsubsection{Fitting the model using fitContinuous}

fitContinuous(phy, dat, SE = 0, 
    model = c("BM","OU","EB","trend","lambda","kappa","delta","drift","white"),
    bounds= list(), control = list(method = c("subplex","L-BFGS-B"),
    niter = 100, FAIL = 1e+200, hessian = FALSE, CI = 0.95), ...)


\subsection{The Ornstein-Uhlenbeck (OU) model}
\subsubsection{The model}

The Ornstein-Uhlenbeck (OU) model (Hansen 1997, Butler and King 2004)describes a mean-reverting process and has the following form, adding an extra term to the Brownian model:
%    dX(t) = -Î±(X(t)-Î¼)+ÏƒdW(t)                          eqn (4)
The parameter %ðœ‡
is a long-term mean, and it is assumed that species evolve around this value. %ð›¼ 
is the strength of evolutionary force that returns traits back towards the mean if they evolve away. The OU model was introduced to population genetics by Lande (1976) to model stabilizing selection in which the mean was recast as a fitness optimum on an adaptive landscape. The process operating in comparative data is analogous, although clearly is not stabilizing selection (despite being sometimes referred to as such). This model has two parameters in addition to those of the Brownian model, %Î± and Î¼.
The OU model predicts that after a time T for a species i, the variance in trait value Xi is:
%var(X_i )=1-e^(-2Î±T)                   eqn (5)
And for a pair of species i and j, the covariance in traits is:
%cov(X_i,X_j )=e^(-2Î±(T-t_ij ) ) (1-e^(-2Î±t_ij ) )                eqn (6)

Put more simply, the Ornstein-Uhlenbeck model, fits a random walk with a central tendency with an attraction strength proportional to the parameter alpha. 

Note that if %Î±
is close to zero then evolution is approximately Brownian.

\subsubsection{Model parameters}

\texttt{fitContinuous} estimates sigma squared, the Brownian rate parameter and alpha, the attraction strength or ``rubber band'' parameter.\\

Default bounds on the Brownian rate parameter are sigsq=c(min=0,max=Inf). Default bounds on alpha = c(min = 0, max = 150)


\subsubsection{Fitting the model using fitContinuous}

fitContinuous(phy, dat, SE = 0, 
    model = c("BM","OU","EB","trend","lambda","kappa","delta","drift","white"),
    bounds= list(), control = list(method = c("subplex","L-BFGS-B"),
    niter = 100, FAIL = 1e+200, hessian = FALSE, CI = 0.95), ...)

\subsection{The Early Burst (EB) model}
\subsubsection{The model}

EB is the Early-burst model (Harmon et al. 2010) and also called the ACDC model (accelerating-decelerating; Blomberg et al. 2003). Set by the a rate parameter, EB fits a model where the rate of evolution increases or decreases exponentially through time, under the model r(t) = r(0) * exp(a * t), where r(0) is the initial rate, a is the rate change parameter, and t is time. The maximum bound is set to -0.000001, representing a decelerating rate of evolution. The minimum bound is set to log(10\^-5)\/depth of the tree.

\subsubsection{Model parameters}

\texttt{fitContinuous} estimates sigma squared, the Brownian rate parameter and a, the rate change parameter. 

Default bounds on the Brownian rate parameter are sigsq=c(min=0,max=Inf). Default bounds on a are -0.000001 for the maximum bound and log(10\^-5)\/depth of the tree for the minimum bound.

\subsubsection{Fitting the model using fitContinuous}

fitContinuous(phy, dat, SE = 0, 
    model = c("BM","OU","EB","trend","lambda","kappa","delta","drift","white"),
    bounds= list(), control = list(method = c("subplex","L-BFGS-B"),
    niter = 100, FAIL = 1e+200, hessian = FALSE, CI = 0.95), ...)


lnl = log likelihood, aic = AIC, aicc = AIC corrected for sample size, k = number of parameters in the model, beta = the Brownian rate parameter/variance, alpha (OU only) = strength of the constraint drawing traits towards the optimum, a (EB only) = early burst parameter . 


  
\subsection{Problems with convergence and bounds}

Above we have mentioned the default bounds on each parameter. Sometimes these need to be changed because the model will not converge. This happens when the likelihood surface has long flat ridges that cause the likelihood search to get ``stuck'' (this is particularly common under the OU model). You can change bounds with the \texttt{bounds} argument in \texttt{fitContinuous}. Several bounds can be given at a time (e.g., bounds=list(sigsq=c(0,0.1),alpha=c(0,1)) would constrain the sigma squared and alpha parameters.

%An example.

\subsection{Comparing among models}

We can use the geiger function fitContinuous to fit the models and can then compare them using AIC (Akaike Information Criterion). You just need to change the model to either â€œBMâ€ for Brownian motion or â€œOUâ€ for OU and â€œEBâ€ for early burst. This is the same function we used to estimate lambda in problem set 1. 

For primate body mass evolution, the model with the smallest AIC is the early burst model, which suggests that an early burst model fits better than either the Brownian motion or OU models. Note that alpha for the OU model is almost zero, and an OU model with an alpha of zero is the same as the Brownian motion model. So we have no evidence for primate body size evolving towards an optimum value.


\end{document}