<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Methods in Macroecology and Macroevolution</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="Practical notes for macro module">
  <meta name="generator" content="bookdown 0.3 and GitBook 2.6.7">

  <meta property="og:title" content="Methods in Macroecology and Macroevolution" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Practical notes for macro module" />
  <meta name="github-repo" content="nhcooper123/TeachingMaterials/MSc_Museum/MacroModule" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Methods in Macroecology and Macroevolution" />
  
  <meta name="twitter:description" content="Practical notes for macro module" />
  

<meta name="author" content="Natalie Cooper (natalie.cooper@nhm.ac.uk)">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="geometric-morphometrics-in-r.html">


<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Macro Methods Module</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Methods in Macroecology and Macroevolution: Module outline</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#module-aims"><i class="fa fa-check"></i><b>1.1</b> Module Aims</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#datasets-and-practical-handouts"><i class="fa fa-check"></i><b>1.2</b> Datasets and practical handouts</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#timetable"><i class="fa fa-check"></i><b>1.3</b> Timetable</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#assessment"><i class="fa fa-check"></i><b>1.4</b> Assessment</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><i class="fa fa-check"></i><b>2</b> What you need to be able to do in R before you start</a><ul>
<li class="chapter" data-level="2.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-r-and-rstudio"><i class="fa fa-check"></i><b>2.1</b> Installing R (and RStudio)</a></li>
<li class="chapter" data-level="2.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#setting-the-working-directory"><i class="fa fa-check"></i><b>2.2</b> Setting the working directory</a></li>
<li class="chapter" data-level="2.3" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#using-a-script"><i class="fa fa-check"></i><b>2.3</b> Using a script</a></li>
<li class="chapter" data-level="2.4" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-and-loading-extra-packages-in-r"><i class="fa fa-check"></i><b>2.4</b> Installing and loading extra packages in R</a></li>
<li class="chapter" data-level="2.5" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-and-viewing-your-data-in-r"><i class="fa fa-check"></i><b>2.5</b> Loading and viewing your data in R</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html"><i class="fa fa-check"></i><b>3</b> Diversity Indices in R</a><ul>
<li class="chapter" data-level="3.1" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#what-are-diversity-indices"><i class="fa fa-check"></i><b>3.1</b> What are diversity indices?</a><ul>
<li class="chapter" data-level="3.1.1" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#alpha-beta-and-gamma-diversity-whittaker-1960-1972"><i class="fa fa-check"></i><b>3.1.1</b> <span class="math inline">\(\alpha\)</span>, <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\gamma\)</span> diversity (Whittaker 1960, 1972)</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#practical-example-using-pokemon"><i class="fa fa-check"></i><b>3.2</b> Practical example using Pokemon…</a></li>
<li class="chapter" data-level="3.3" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#species-diversity-indices"><i class="fa fa-check"></i><b>3.3</b> Species diversity indices</a><ul>
<li class="chapter" data-level="3.3.1" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#alpha-diversity"><i class="fa fa-check"></i><b>3.3.1</b> <span class="math inline">\(\alpha\)</span> diversity</a></li>
<li class="chapter" data-level="3.3.2" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#beta-diversity"><i class="fa fa-check"></i><b>3.3.2</b> <span class="math inline">\(\beta\)</span> diversity</a></li>
<li class="chapter" data-level="3.3.3" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#gamma-diversity"><i class="fa fa-check"></i><b>3.3.3</b> <span class="math inline">\(\gamma\)</span> diversity</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#species-accumulation-curves-colwell-coddington-1994"><i class="fa fa-check"></i><b>3.4</b> Species accumulation curves (Colwell &amp; Coddington 1994)</a><ul>
<li class="chapter" data-level="3.4.1" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#practical-example"><i class="fa fa-check"></i><b>3.4.1</b> Practical example</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#references"><i class="fa fa-check"></i><b>3.5</b> References</a><ul>
<li class="chapter" data-level="3.5.1" data-path="diversity-indices-in-r.html"><a href="diversity-indices-in-r.html#extra-reading"><i class="fa fa-check"></i><b>3.5.1</b> Extra reading</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html"><i class="fa fa-check"></i><b>4</b> Visualising phylogenies in R</a><ul>
<li class="chapter" data-level="4.1" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#loading-your-phylogeny-and-data-into-r"><i class="fa fa-check"></i><b>4.1</b> Loading your phylogeny and data into R</a><ul>
<li class="chapter" data-level="4.1.1" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#reading-in-a-phylogeny-from-a-file"><i class="fa fa-check"></i><b>4.1.1</b> Reading in a phylogeny from a file</a></li>
<li class="chapter" data-level="4.1.2" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#reading-in-a-phylogeny-that-is-already-built-into-r"><i class="fa fa-check"></i><b>4.1.2</b> Reading in a phylogeny that is already built into R</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#basic-tree-viewing-in-r"><i class="fa fa-check"></i><b>4.2</b> Basic tree viewing in R</a></li>
<li class="chapter" data-level="4.3" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#adding-trait-data-to-trees-in-r"><i class="fa fa-check"></i><b>4.3</b> Adding trait data to trees in R</a><ul>
<li class="chapter" data-level="4.3.1" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#ancestral-state-reconstructions-on-continuous-data"><i class="fa fa-check"></i><b>4.3.1</b> Ancestral state reconstructions on continuous data</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#phylomorphospace-plots-in-r"><i class="fa fa-check"></i><b>4.4</b> Phylomorphospace plots in R</a></li>
<li class="chapter" data-level="4.5" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#phylogenies-and-maps"><i class="fa fa-check"></i><b>4.5</b> Phylogenies and maps!</a></li>
<li class="chapter" data-level="4.6" data-path="visualising-phylogenies-in-r.html"><a href="visualising-phylogenies-in-r.html#a-new-package-for-visualising-trees-ggtree"><i class="fa fa-check"></i><b>4.6</b> A new package for visualising trees: ggtree</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html"><i class="fa fa-check"></i><b>5</b> Phylogenetic Generalised Least Squares (PGLS) in R</a><ul>
<li class="chapter" data-level="5.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#preparing-for-the-analysis"><i class="fa fa-check"></i><b>5.1</b> Preparing for the analysis</a><ul>
<li class="chapter" data-level="5.1.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#load-the-required-packages"><i class="fa fa-check"></i><b>5.1.1</b> Load the required packages</a></li>
<li class="chapter" data-level="5.1.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#reading-and-checking-your-data-in-r"><i class="fa fa-check"></i><b>5.1.2</b> Reading and checking your data in R</a></li>
<li class="chapter" data-level="5.1.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#reading-and-checking-your-phylogeny-in-r"><i class="fa fa-check"></i><b>5.1.3</b> Reading and checking your phylogeny in R</a></li>
<li class="chapter" data-level="5.1.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#manipulating-your-data-and-phylogeny-in-r"><i class="fa fa-check"></i><b>5.1.4</b> Manipulating your data and phylogeny in R</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#ordinary-least-squares-ols-regression"><i class="fa fa-check"></i><b>5.2</b> Ordinary least squares (OLS) regression</a></li>
<li class="chapter" data-level="5.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#phylogenetic-generalized-least-squares-models-pgls"><i class="fa fa-check"></i><b>5.3</b> Phylogenetic generalized least squares models (PGLS)</a><ul>
<li class="chapter" data-level="5.3.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#likelihood-profiles-for-lambda-in-pgls-models"><i class="fa fa-check"></i><b>5.3.1</b> Likelihood profiles for <span class="math inline">\(\lambda\)</span> in PGLS models</a></li>
<li class="chapter" data-level="5.3.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#model-diagnostics-for-pgls-models"><i class="fa fa-check"></i><b>5.3.2</b> Model diagnostics for PGLS models</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#estimating-phylogenetic-signal-for-one-variable"><i class="fa fa-check"></i><b>5.4</b> Estimating phylogenetic signal for one variable</a><ul>
<li class="chapter" data-level="5.4.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#pagels-lambda-pagel-19971999-freckleton-et-al-2002"><i class="fa fa-check"></i><b>5.4.1</b> Pagel’s <span class="math inline">\(\lambda\)</span> (Pagel 1997/1999, Freckleton et al 2002)</a></li>
<li class="chapter" data-level="5.4.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#blombergs-k-blomberg-et-al-2003"><i class="fa fa-check"></i><b>5.4.2</b> Blomberg’s K (Blomberg et al 2003)</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#references-1"><i class="fa fa-check"></i><b>5.5</b> References</a><ul>
<li class="chapter" data-level="5.5.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#extra-reading-1"><i class="fa fa-check"></i><b>5.5.1</b> Extra Reading</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html"><i class="fa fa-check"></i><b>6</b> Macroevolutionary models in R: Part 1 - continuous traits</a><ul>
<li class="chapter" data-level="6.1" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#preparing-for-the-analysis-1"><i class="fa fa-check"></i><b>6.1</b> Preparing for the analysis</a><ul>
<li class="chapter" data-level="6.1.1" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#load-packages-read-in-the-data-and-the-tree"><i class="fa fa-check"></i><b>6.1.1</b> Load packages, read in the data and the tree</a></li>
<li class="chapter" data-level="6.1.2" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#modify-the-tree-and-data-so-they-can-be-used-in-the-analyses."><i class="fa fa-check"></i><b>6.1.2</b> Modify the tree and data so they can be used in the analyses.</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#models-of-continuous-trait-evolution"><i class="fa fa-check"></i><b>6.2</b> Models of continuous trait evolution</a><ul>
<li class="chapter" data-level="6.2.1" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#the-brownian-motion-bm-model"><i class="fa fa-check"></i><b>6.2.1</b> The Brownian motion (BM model)</a></li>
<li class="chapter" data-level="6.2.2" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#the-ornstein-uhlenbeck-ou-model"><i class="fa fa-check"></i><b>6.2.2</b> The Ornstein-Uhlenbeck (OU) model</a></li>
<li class="chapter" data-level="6.2.3" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#the-early-burst-eb-model"><i class="fa fa-check"></i><b>6.2.3</b> The Early Burst (EB) model</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#fitting-the-models-using-fitcontinuous"><i class="fa fa-check"></i><b>6.3</b> Fitting the models using <code>fitContinuous</code></a><ul>
<li class="chapter" data-level="6.3.1" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#comparing-models-of-evolution"><i class="fa fa-check"></i><b>6.3.1</b> Comparing models of evolution</a></li>
<li class="chapter" data-level="6.3.2" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#problems-with-convergence-and-bounds"><i class="fa fa-check"></i><b>6.3.2</b> Problems with convergence and bounds</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="macroevolutionary-models-in-r-part-1-continuous-traits.html"><a href="macroevolutionary-models-in-r-part-1-continuous-traits.html#references-2"><i class="fa fa-check"></i><b>6.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html"><i class="fa fa-check"></i><b>7</b> Macroevolutionary models in R: Part 2 - discrete traits</a><ul>
<li class="chapter" data-level="7.1" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#preparing-for-the-analysis-2"><i class="fa fa-check"></i><b>7.1</b> Preparing for the analysis</a><ul>
<li class="chapter" data-level="7.1.1" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#load-packages-read-in-the-data-and-the-tree-1"><i class="fa fa-check"></i><b>7.1.1</b> Load packages, read in the data and the tree</a></li>
<li class="chapter" data-level="7.1.2" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#modify-the-tree-and-data-so-they-can-be-used-in-the-analyses.-1"><i class="fa fa-check"></i><b>7.1.2</b> Modify the tree and data so they can be used in the analyses.</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#fitting-models-of-evolution-to-discrete-data-regime-dependent-evolution"><i class="fa fa-check"></i><b>7.2</b> Fitting models of evolution to discrete data (regime dependent evolution)</a></li>
<li class="chapter" data-level="7.3" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#models-of-discrete-trait-evolution"><i class="fa fa-check"></i><b>7.3</b> Models of discrete trait evolution</a><ul>
<li class="chapter" data-level="7.3.1" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#mk1-all-rates-equal-er"><i class="fa fa-check"></i><b>7.3.1</b> Mk1 – all rates equal (ER)</a></li>
<li class="chapter" data-level="7.3.2" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#mk-symmetric-rates-sym"><i class="fa fa-check"></i><b>7.3.2</b> Mk – symmetric rates (SYM)</a></li>
<li class="chapter" data-level="7.3.3" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#mk-all-rates-different-ard"><i class="fa fa-check"></i><b>7.3.3</b> Mk – All Rates Different (ARD)</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#fitting-the-models-using-fitdiscrete"><i class="fa fa-check"></i><b>7.4</b> Fitting the models using <code>fitDiscrete</code></a></li>
<li class="chapter" data-level="7.5" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#ancestral-state-reconstructions"><i class="fa fa-check"></i><b>7.5</b> Ancestral state reconstructions</a></li>
<li class="chapter" data-level="7.6" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#fitting-models-using-ouwie"><i class="fa fa-check"></i><b>7.6</b> Fitting models using <code>OUwie</code></a><ul>
<li class="chapter" data-level="7.6.1" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#multi-rate-brownian-motion-bms"><i class="fa fa-check"></i><b>7.6.1</b> Multi-rate Brownian motion (BMS)</a></li>
<li class="chapter" data-level="7.6.2" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#multi-peak-ou-models"><i class="fa fa-check"></i><b>7.6.2</b> Multi-peak OU models</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="macroevolutionary-models-in-r-part-2-discrete-traits.html"><a href="macroevolutionary-models-in-r-part-2-discrete-traits.html#references-3"><i class="fa fa-check"></i><b>7.7</b> References</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html"><i class="fa fa-check"></i><b>8</b> Geometric Morphometrics in R</a><ul>
<li class="chapter" data-level="8.1" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#a-quick-introduction-to-geometric-morphometrics"><i class="fa fa-check"></i><b>8.1</b> A quick introduction to geometric morphometrics</a><ul>
<li class="chapter" data-level="8.1.1" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#what-is-morphometrics"><i class="fa fa-check"></i><b>8.1.1</b> What is morphometrics?</a></li>
<li class="chapter" data-level="8.1.2" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#landmarks"><i class="fa fa-check"></i><b>8.1.2</b> Landmarks</a></li>
<li class="chapter" data-level="8.1.3" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#collecting-data-and-adding-landmarks"><i class="fa fa-check"></i><b>8.1.3</b> Collecting data and adding landmarks</a></li>
<li class="chapter" data-level="8.1.4" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#measurement-error"><i class="fa fa-check"></i><b>8.1.4</b> Measurement error</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#a-practical-example-using-r"><i class="fa fa-check"></i><b>8.2</b> A practical example using R</a></li>
<li class="chapter" data-level="8.3" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#gathering-the-data"><i class="fa fa-check"></i><b>8.3</b> Gathering the data</a></li>
<li class="chapter" data-level="8.4" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#digitising-adding-landmarks-and-curves"><i class="fa fa-check"></i><b>8.4</b> Digitising (adding landmarks and curves)</a></li>
<li class="chapter" data-level="8.5" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#reading-in-your-data-and-plotting-it"><i class="fa fa-check"></i><b>8.5</b> Reading in your data and plotting it</a></li>
<li class="chapter" data-level="8.6" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#generalised-procrustes-superimposition-gpa"><i class="fa fa-check"></i><b>8.6</b> Generalised Procrustes Superimposition (GPA)</a></li>
<li class="chapter" data-level="8.7" data-path="geometric-morphometrics-in-r.html"><a href="geometric-morphometrics-in-r.html#principal-components-and-plotting"><i class="fa fa-check"></i><b>8.7</b> Principal components and plotting</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html"><i class="fa fa-check"></i><b>9</b> Statistical analyses of geometric morphometric datasets</a><ul>
<li class="chapter" data-level="9.1" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#regression"><i class="fa fa-check"></i><b>9.1</b> Regression</a></li>
<li class="chapter" data-level="9.2" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#multivariate-regression"><i class="fa fa-check"></i><b>9.2</b> Multivariate regression</a></li>
<li class="chapter" data-level="9.3" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#anova"><i class="fa fa-check"></i><b>9.3</b> ANOVA</a></li>
<li class="chapter" data-level="9.4" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#manova"><i class="fa fa-check"></i><b>9.4</b> MANOVA</a></li>
<li class="chapter" data-level="9.5" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#resources-for-learning-geometric-morphometrics"><i class="fa fa-check"></i><b>9.5</b> Resources for learning geometric morphometrics</a></li>
<li class="chapter" data-level="9.6" data-path="statistical-analyses-of-geometric-morphometric-datasets.html"><a href="statistical-analyses-of-geometric-morphometric-datasets.html#references-4"><i class="fa fa-check"></i><b>9.6</b> References</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/nhcooper123/TeachingMaterials/MSc_Museum/MacroModule" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Methods in Macroecology and Macroevolution</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="statistical-analyses-of-geometric-morphometric-datasets" class="section level1">
<h1><span class="header-section-number">Practical 9</span> Statistical analyses of geometric morphometric datasets</h1>
<p>This is the point at which geometric morphometrics gets really exciting - it’s the point that you get to answer whatever question you started out with! It’s also the point at which the number of different options becomes very large. Two common analyses are regressions and multivariate ANOVA (MANOVA). Regressions are used when you have continuous explanatory variables, for example if you want to see if body size is correlated with shape. MANOVA (or ANOVA) is used when you have categorical explanatory variables, for example if you want to see if habitat type is correlated with shape.</p>
<p>At this stage of the analysis we often add a new dataset containing our variables of interest. For simplicity we will just invent some data as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create new dataframe with PC scores, body size and diet.</span>
<span class="co"># Body sizes from a random normal distribution with mean of 100 and sd of 25.</span>
<span class="co"># Diet assigned as squid to first two specimens and fish to others.</span>
mydata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(pca.landmarks$pc.scores, 
                     <span class="dt">body.size =</span> <span class="kw">rnorm</span>(<span class="dt">n =</span> <span class="dv">8</span>, <span class="dt">mean =</span> <span class="dv">100</span>, <span class="dt">sd =</span> <span class="dv">25</span>), 
                     <span class="dt">diet =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;squid&quot;</span>,<span class="dv">2</span>), <span class="kw">rep</span>(<span class="st">&quot;fish&quot;</span>, <span class="dv">6</span>)) 
                     )
<span class="co"># Look at the data</span>
<span class="kw">str</span>(mydata)</code></pre></div>
<pre><code>## &#39;data.frame&#39;:    8 obs. of  10 variables:
##  $ PC1      : num  -0.2228 0.0468 0.0258 0.0349 0.125 ...
##  $ PC2      : num  -0.0499 0.00149 0.02146 -0.04968 -0.07385 ...
##  $ PC3      : num  -0.000638 -0.041751 0.016942 0.000896 0.011896 ...
##  $ PC4      : num  -0.00103 0.00981 0.03057 0.00119 -0.01084 ...
##  $ PC5      : num  0.0068 0.00444 0.00488 -0.02318 0.00956 ...
##  $ PC6      : num  -7.16e-05 -8.64e-03 5.17e-03 -2.21e-03 4.33e-03 ...
##  $ PC7      : num  0.000387 -0.003641 0.001733 0.004996 -0.003752 ...
##  $ PC8      : num  0.00 2.08e-17 2.06e-17 1.39e-17 2.99e-17 ...
##  $ body.size: num  114.9 93.8 81.8 92 154.6 ...
##  $ diet     : Factor w/ 2 levels &quot;fish&quot;,&quot;squid&quot;: 2 2 1 1 1 1 1 1</code></pre>
<div id="regression" class="section level2">
<h2><span class="header-section-number">9.1</span> Regression</h2>
<p>Regressions are done in the same way as usual.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit model</span>
model1 &lt;-<span class="st"> </span><span class="kw">lm</span>(PC1 ~<span class="st"> </span>body.size, <span class="dt">data =</span> mydata)

<span class="co"># Assess assumptions</span>
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(model1)</code></pre></div>
<p><img src="macro-module_files/figure-html/unnamed-chunk-155-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))

<span class="co"># Look at overall model significance</span>
<span class="kw">anova</span>(model1)</code></pre></div>
<pre><code>## Analysis of Variance Table
## 
## Response: PC1
##           Df   Sum Sq   Mean Sq F value Pr(&gt;F)
## body.size  1 0.000762 0.0007623  0.0643 0.8083
## Residuals  6 0.071104 0.0118506</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at parameters and their significance</span>
<span class="kw">summary</span>(model1)</code></pre></div>
<pre><code>## 
## Call:
## lm(formula = PC1 ~ body.size, data = mydata)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.22705 -0.02121  0.03891  0.04604  0.10261 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)
## (Intercept) -0.048240   0.194059  -0.249    0.812
## body.size    0.000457   0.001802   0.254    0.808
## 
## Residual standard error: 0.1089 on 6 degrees of freedom
## Multiple R-squared:  0.01061,    Adjusted R-squared:  -0.1543 
## F-statistic: 0.06433 on 1 and 6 DF,  p-value: 0.8083</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot</span>
<span class="kw">plot</span>(PC1 ~<span class="st"> </span>body.size, <span class="dt">data =</span> mydata, <span class="dt">pch =</span> <span class="dv">16</span>)
<span class="kw">abline</span>(model1)</code></pre></div>
<p><img src="macro-module_files/figure-html/unnamed-chunk-155-2.png" width="672" /></p>
<ul>
<li>Can you remember how to interpret these results?</li>
<li>Why will your results be different to mine?</li>
</ul>
</div>
<div id="multivariate-regression" class="section level2">
<h2><span class="header-section-number">9.2</span> Multivariate regression</h2>
<p>Multivariate regression is the same as univariate regression but is used where you have multiple response variables, i.e. if we wanted to include PC1 and PC2 in the analysis. It’s quite easy to implement, we just use <code>cbind</code> to bind together the different variables we want to include. Note that assumption checking isn’t possible, nor is plotting in the usual way. Additionally, <code>summary(model)</code> presents the parameters for each response variable separately (e.g. <code>PC1 ~ body.size</code> then <code>PC2 ~ body.size</code>) so is not particularly useful here.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit model</span>
model2 &lt;-<span class="st"> </span><span class="kw">lm</span>(<span class="kw">cbind</span>(mydata$PC1,mydata$PC2) ~<span class="st"> </span>body.size, <span class="dt">data =</span> mydata)

<span class="co"># Look at overall model significance</span>
<span class="kw">anova</span>(model2)</code></pre></div>
<pre><code>## Analysis of Variance Table
## 
##             Df  Pillai approx F num Df den Df Pr(&gt;F)
## (Intercept)  1 0.00000   0.0000      2      5 1.0000
## body.size    1 0.28641   1.0034      2      5 0.4302
## Residuals    6</code></pre>
<ul>
<li>Can you guess how to interpret these results given what you know about regression?</li>
</ul>
</div>
<div id="anova" class="section level2">
<h2><span class="header-section-number">9.3</span> ANOVA</h2>
<p>ANOVAs are done in the same way as usual.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit model</span>
model3 &lt;-<span class="st"> </span><span class="kw">lm</span>(PC1 ~<span class="st"> </span>diet, <span class="dt">data =</span> mydata)

<span class="co"># Assess assumptions</span>
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(model3)</code></pre></div>
<p><img src="macro-module_files/figure-html/unnamed-chunk-157-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))

<span class="co"># Look at overall model significance</span>
<span class="kw">anova</span>(model3)</code></pre></div>
<pre><code>## Analysis of Variance Table
## 
## Response: PC1
##           Df   Sum Sq  Mean Sq F value Pr(&gt;F)
## diet       1 0.020644 0.020644  2.4182 0.1709
## Residuals  6 0.051222 0.008537</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at parameters and their significance</span>
<span class="kw">summary</span>(model3)</code></pre></div>
<pre><code>## 
## Call:
## lm(formula = PC1 ~ diet, data = mydata)
## 
## Residuals:
##       Min        1Q    Median        3Q       Max 
## -0.134813 -0.049412  0.001039  0.029529  0.134813 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)
## (Intercept)  0.02933    0.03772   0.778    0.466
## dietsquid   -0.11731    0.07544  -1.555    0.171
## 
## Residual standard error: 0.0924 on 6 degrees of freedom
## Multiple R-squared:  0.2873, Adjusted R-squared:  0.1685 
## F-statistic: 2.418 on 1 and 6 DF,  p-value: 0.1709</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot</span>
<span class="kw">plot</span>(PC1 ~<span class="st"> </span>diet, <span class="dt">data =</span> mydata, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;springgreen&quot;</span>))</code></pre></div>
<p><img src="macro-module_files/figure-html/unnamed-chunk-157-2.png" width="672" /></p>
<ul>
<li>Can you remember how to interpret these results?</li>
</ul>
</div>
<div id="manova" class="section level2">
<h2><span class="header-section-number">9.4</span> MANOVA</h2>
<p>MANOVA is the same as ANOVA but is used where you have multiple response variables, i.e. if we wanted to include PC1 and PC2 in the analysis. It’s quite easy to implement, we just use <code>cbind</code> to bind together the different variables we want to include. Note that assumption checking isn’t possible for MANOVA in R at this time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit model</span>
model4 &lt;-<span class="st"> </span><span class="kw">manova</span>(<span class="kw">cbind</span>(mydata$PC1,mydata$PC2) ~<span class="st"> </span>diet, <span class="dt">data =</span> mydata)

<span class="co"># Look at overall model significance</span>
<span class="kw">anova</span>(model4)</code></pre></div>
<pre><code>## Analysis of Variance Table
## 
##             Df  Pillai approx F num Df den Df Pr(&gt;F)
## (Intercept)  1 0.00000   0.0000      2      5 1.0000
## diet         1 0.37191   1.4803      2      5 0.3127
## Residuals    6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at parameters and their significance</span>
<span class="kw">summary</span>(model4)</code></pre></div>
<pre><code>##           Df  Pillai approx F num Df den Df Pr(&gt;F)
## diet       1 0.37191   1.4803      2      5 0.3127
## Residuals  6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot </span>
<span class="kw">plot</span>(<span class="kw">cbind</span>(mydata$PC1,mydata$PC2) ~<span class="st"> </span>diet, <span class="dt">data =</span> mydata, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;springgreen&quot;</span>))</code></pre></div>
<p><img src="macro-module_files/figure-html/unnamed-chunk-158-1.png" width="672" /></p>
<ul>
<li>Can you guess how to interpret these results given what you know about ANOVA?</li>
</ul>
<p>Of course the results here are nonsensical as we made u pthe data, but hopefully this gives you an idea of how we might use geometric morphometrics data in analyses. Other common analyses look at disparity of groups, convergence and divergence, evolution of shape and shape space etc.</p>
<hr />
</div>
<div id="resources-for-learning-geometric-morphometrics" class="section level2">
<h2><span class="header-section-number">9.5</span> Resources for learning geometric morphometrics</h2>
<ol style="list-style-type: decimal">
<li><p>I highly advise getting hold of this book (<a href="http://store.elsevier.com/Geometric-Morphometrics-for-Biologists/Miriam-Zelditch/isbn-9780123869036/">Zelditch et al. 2012</a>) from the library and reading the first few chapters, plus any chapters later in the book that are relevant to the analyses you will be doing. Don’t panic too much about the maths or the equations, just try to get a general understanding of what each method is doing, especially GPA and PCA.</p></li>
<li><p>Another useful book is <a href="http://lib.du.ac.ir/documents/10157/60743/Morphometrics+With+R.pdf">Morphometrics with R</a> by Julien Claude. It is a little harder to read than Zelditch, but more focused on practical analysis in R. Note that I’d generally advise using the <a href="https://cran.r-project.org/web/packages/geomorph/geomorph.pdf">geomorph</a> package (see links below) to do these analyses in R, but many of the principles are the same in this book which uses other methods. It’s available as a <a href="http://lib.du.ac.ir/documents/10157/60743/Morphometrics+With+R.pdf">PDF</a>.</p></li>
<li><p>David Polly has an excellent <a href="http://www.indiana.edu/~g562/">set of lectures</a> about all basic topics in geometric morphometrics including PCA and GPA. Note that these use Mathematica not R. There are also slides <a href="http://www.indiana.edu/~g562/PBDB2013/">here</a> from an R based course with a basic workflow for an <a href="http://www.indiana.edu/~g562/PBDB2013/Day%202B%20-%20Geometric%20Morphometrics%20in%20R.pdf">analysis using geomorph</a>. I’d recommend starting any project with a quick flick through these <a href="http://www.indiana.edu/~g562/PBDB2013/Day%202A%20-%20Introduction%20to%20Geometric%20Morphometrics.pdf">intro slides</a></p></li>
<li><p>Emma Sherratt has put together an excellent tutorial/vignette for <a href="http://www.public.iastate.edu/~dcadams/PDFPubs/Quick%20Guide%20to%20Geomorph%20v2.0.pdf">geomorph</a> from inputting landmarks to complex analyses.</p></li>
<li><p>The <a href="https://cran.r-project.org/web/packages/geomorph/geomorph.pdf">geomorph vignette</a> may also be helpful for more complex analyses.</p></li>
<li><p>For error checking, take a look at <a href="http://link.springer.com/article/10.1007/s00427-016-0537-4">Fruciano 2016</a>, a review of the subject in Evolution. Also take a look at <a href="http://lib.du.ac.ir/documents/10157/60743/Morphometrics+With+R.pdf">Claude</a> pages 63-65 for ideas on sources of error.</p></li>
</ol>
</div>
<div id="references-4" class="section level2">
<h2><span class="header-section-number">9.6</span> References</h2>
<ul>
<li>Zelditch, M.L., Swiderski, D.L., and Sheets, H.D.. 2012. Geometric Morphometrics for Biologists: A Primer. Academic Press.</li>
<li>Claude, Julien. Morphometrics with R. Springer Science &amp; Business Media, 2008.</li>
<li>Fruciano, C. Measurement error in geometric morphometrics. 2016. Development genes and evolution 226:139-158.</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="geometric-morphometrics-in-r.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>


<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/nhcooper123/TeachingMaterials/MSc_Museum/MacroModule/edit/master/7_GeometricMorphometricsinR.Rmd",
"text": "Edit"
},
"download": ["macro-module.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
